Add debug module for measuring CPU load

nordic-krch opened the pull request:
Content:
- cpu_load module with API for retrieving CPU load and shell commands
- test

Module is using one TIMER instance and SLEEPENTER SLEEPEXIT events (2 PPI channels) to accurately measure CPU load.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:72: ERROR:TRAILING_WHITESPACE: trailing whitespace
#72: FILE: include/debug/cpu_load.rst:13:
+ $

-:134: ERROR:TRAILING_WHITESPACE: trailing whitespace
#134: FILE: subsys/debug/cpu_load/Kconfig:15:
+^I$

-:259: ERROR:SPACING: space required before the open brace '{'
#259: FILE: subsys/debug/cpu_load/cpu_load.c:75:
+void timer_handler(nrf_timer_event_t event_type, void *context){}

-:267: WARNING:LINE_SPACING: Missing a blank line after declarations
#267: FILE: subsys/debug/cpu_load/cpu_load.c:83:
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	config.frequency = NRF_TIMER_FREQ_1MHz;

-:333: ERROR:SPACING: need consistent spacing around '*' (ctx:WxV)
#333: FILE: subsys/debug/cpu_load/cpu_load.c:149:
+			(10000 *(total_us - sleep_us))/total_us : 0;
 			       ^

- total: 4 errors, 1 warnings, 401 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

## Codeowners issues

```
New files added that are not covered in CODEOWNERS:

subsys/debug/cpu_load/CMakeLists.txt
subsys/debug/cpu_load/cpu_load.c

Please add one or more entries in the CODEOWNERS file to cover those files
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

nordic-krch commented:
Interesting observation. After enabling cpu_load in zephyr shell sample the load for pca10056 and pca10090 in idle significantly differs: 
- pca10056 - 0,45%
- pca10090  - 1,74%

pdunaj commented:
Hi @nordic-krch , there is a module for CPU load measurement that uses profiling events (see CONFIG_TRACING_CPU_STATS). Do you think there is a point in having two? Maybe the one we have now should be removed (since profiling is intrusive it should have lower accuracy)?

BTW - could you add a similar option as the other one has - to enable periodic printing of the CPU usage without shell? Shell is large and it is in some cases impractical to enable it.

MarekPieta commented:
This implementation can only be used on nrf. The solution based on profiling hooks may still be useful for other microcontrollers.
Periodic displaying CPU usage (e.g. using `printk`) would be nice feature. This could be done in similar way as in the module based on profiling hooks.

nordic-krch commented:
i tried to find it and failed. Seems that i didn't search for `stats`. trace stats is generic but more intrusive and a bit less accurate since it does not take into account out of kernel WFE or WFI calls. In the system with frequent interrupts trace overhead (`sys_trace_isr_enter`/`sys_trace_isr_exit`) may be significant. What do you think? I can close this PR or refer to trace in the cpu_load documention explaining the differences.

pdunaj commented:
> i tried to find it and failed. Seems that i didn't search for `stats`. trace stats is generic but more intrusive and a bit less accurate since it does not take into account out of kernel WFE or WFI calls. In the system with frequent interrupts trace overhead (`sys_trace_isr_enter`/`sys_trace_isr_exit`) may be significant. What do you think? I can close this PR or refer to trace in the cpu_load documention explaining the differences.

Why closing this PR? I think the functionality would be very useful if values can be provided in less intrusive way. If anything I would thught about removing the old CPU load measurement method. @mapi is however right - the old way would still be useful on the other SoCs.

BTW - I am wondering what are the measured loads if you enable both.

MarekPieta commented:
@nordic-krch, I just wanted not to remove the old implementation - it still could be used for other SoC's.
Your method is way less intrusive, so it would be better for nrf SoC.

nordic-krch commented:
> BTW - I am wondering what are the measured loads if you enable both.

I prepared a sample with frequent interrupt from TIMER. CPU load reported by `cpu_load` (let say non-intrusive) reported 64,5%. When I enabled `log_stats` but without printing because that would impact measurement, cpu load reported by `cpu_load` module was 73,6%. I seems that in case of frequent interrupts the overhead of `cpu_stats` is significant.

 > Periodic displaying CPU usage (e.g. using printk) would be nice feature. 

Added.

I've also added an option to drive TIMER (which measures sleep period) using RTC1 tick event. This option has 2 advantages: 
- same clock source is used for sleep and active period measurement so the result is more accurate (if RC is used as clock source) then HF and LF frequencies are not exactly aligned.
- HF clock is not required for sleep measurement so it allows measuring cpu load and going to low power idle state
That comes with the cost of 1 additional PPI channel.

nordic-krch commented:
> Interesting observation. After enabling cpu_load in zephyr shell sample the load for pca10056 and pca10090 in idle significantly differs

Fyi, reason for that was discrepancy between HF and LF accuracy on pca10090.

lemrey commented:
@b-gent could you please have a look at the documentation?

greg-fer commented:
I'll be reviewing this instead of @b-gent .

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+ * @{
+ */
+
+/** @brief Initialize CPU load measurement module.
```
```suggestion
/** @brief Initialize the CPU load measurement module.
```

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+
+/** @brief Initialize CPU load measurement module.
+ *
+ * During initialization TIMER driver and PPI channels are allocated.
```
```suggestion
 * The TIMER driver and PPI channels are allocated during the initialization of this module.
```

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+ *
+ * During initialization TIMER driver and PPI channels are allocated.
+ *
+ * @return 0 on success, error code otherwise.
```
```suggestion
* @retval 0 The initialization is successful.
* @retval error The initialization was not successful. See the error code for the specific reason.
```

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+ */
+int cpu_load_init(void);
+
+/** @brief Reset measurement. */
```
```suggestion
/** @brief Reset the CPU load measurement module. */
```
I guess this can be called only after intialization? Can we mention it here in the documentation? Are the CPU load measurement values reset a spart of this function? Are the previous values stored somewhere on reset?

nordic-krch replied:
its not module reset but measurement reset. So suggestion is not valid.

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+/** @brief Reset measurement. */
+void cpu_load_reset(void);
+
+/** @brief Get CPU load.
```
```suggestion
/** @brief Get the CPU load measurement value.
```

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+
+/** @brief Get CPU load.
+ *
+ * Load is represented in 0,001% units where 100000 value represents 100% load.
```
```suggestion
* The CPU load is represented in 0,001% units where a 100000 value represents 100% load.
```
Can you add an example value here? Where do you get these values? In the shell?

greg-fer commented on include/debug/cpu_load.h:
```diff
@@ -0,0 +1,47 @@
+ *
+ * Load is represented in 0,001% units where 100000 value represents 100% load.
+ *
+ * @return CPU load.
```
```suggestion
 * @return The current CPU load value.
```

greg-fer commented on subsys/debug/cpu_load/Kconfig:
```diff
@@ -0,0 +1,82 @@
+config CPU_LOAD_ALIGNED_CLOCKS
+	bool "Enable aligned clock sources"
+	help
+	  If enabled then sleep period measurement is done using the same clock
```
```
After enabling this option, the sleep period measurement
is done using the same clock source as the active period,
and one additional PPI channel is used.
If you use the internal RC oscillator for at least one
of the clock sources (high or low frequency), this option
will ensure more accurate results.
Enabling this option allows going to low power idle mode
because the high frequency clock is not used by this module.
```

greg-fer commented on subsys/debug/cpu_load/Kconfig:
```diff
@@ -0,0 +1,82 @@
+	select NRFX_DPPI if HAS_HW_NRF_DPPIC
+	depends on !SOC_SERIES_NRF51X #Lack of required HW events
+	help
+	  Enable CPU load measurement instrumentation. Tool is using one TIMER
```
```
Enable the CPU load measurement instrumentation.
This tool is using one TIMER peripheral and PPI
to perform accurate CPU load measurement.
```

MarekPieta commented:
Do you know when this will be merged?
We would like to use it to profile CPU usage in `nrf_desktop` project.

nordic-krch commented:
@thomasstenersen  any idea what is that?

```
error: unable to write file third_party/mbedtls/repo/tests/data_files/hash_file_5

error: unable to write file third_party/mbedtls/repo/tests/data_files/keyUsage.decipherOnly.crt

error: unable to write file third_party/mbedtls/repo/tests/data_files/mpi_10

error: unable to write file third_party/mbedtls/repo/tests/data_files/mpi_too_big

error: unable to write file third_party/mbedtls/repo/tests/data_files/passwd.psk

error: unable to write file third_party/mbedtls/repo/tests/data_files/print_c.pl

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa4096_prv.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa4096_pub.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa521.key

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa522.key

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa528.key

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_3des.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_aes128.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_aes192.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_aes256.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_clear.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_1024_des.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_3des.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_aes128.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_aes192.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_aes256.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_clear.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_des.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_public.der

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_2048_public.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_4096_3des.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_4096_aes128.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_4096_aes192.pem

error: unable to write file third_party/mbedtls/repo/tests/data_files/rsa_pkcs1_4096_aes256.pem

error: unable to write file third_party/mbedtls/repo/testwrapper script does not seem to be touching the log file in /jenkins_workspace/workspace/latest_fw-nrfconnect-nrf_PR-1272@2@tmp/durable-55194ad4

(JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)

script returned exit code -1
```

nordic-krch commented:
@MarekPieta  @anangl  could you review?

thomasstenersen commented:
@nordic-krch I guess you meant to tag @thst-nordic  :) I don't have any idea about that log snippet :(

nordic-krch commented:
@thomasstenersen right, sorry I'm rarely working with ncs and mixed names. @thst-nordic can you help?

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,234 @@
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		/* ppi free */
```
Was that (freeing of the allocated channels) planned to be added?

nordic-krch replied:
added ppi cleanup which frees channels. However, with DPPI it is not that simple as cpu_load supports a case when event is already allocated and then it just subscribes to it. I added it as an option (and extended test to check that).

I also had to take into account this when enabling ppi channels. "shared" channels are not enabled.

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,234 @@
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
```
Shouldn't it be placed after the check of `ready`? Currently the PPI channel for the TICK event is allocated on each call to `cpu_load_init`.

SebastianBoe commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,234 @@
+/* Define to please compiler when periodic logging is disabled. */
+#ifndef CONFIG_CPU_LOAD_LOG_INTERVAL
+#define CONFIG_CPU_LOAD_LOG_INTERVAL 0
+#endif
```
It is not allowed to define macros that are in the Kconfig namespace, like CONFIG_CPU_LOAD_LOG_INTERVAL.

You can either have CONFIG_CPU_LOAD_LOG_INTERVAL not depend on PERIODIC, and have the default value be 0 when not PERIODIC.

Or instead of referencing CONFIG_CPU_LOAD_LOG_INTERVAL in code, you can reference CPU_LOAD_LOG_INTERVAL.

nordic-krch replied:
> Or instead of referencing CONFIG_CPU_LOAD_LOG_INTERVAL in code, you can reference CPU_LOAD_LOG_INTERVAL.

You mean something like this?
```
#ifdef CONFIG_CPU_LOAD_LOG_INTERVAL 
#define CPU_LOAD_LOG_INTERVAL CONFIG_CPU_LOAD_LOG_INTERVAL 
#else 
#define CPU_LOAD_LOG_INTERVAL 0
#else 
```

SebastianBoe replied:
yes

nordic-krch replied:
done

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,234 @@
+{
+	nrfx_timer_enable(&timer);
+	nrfx_timer_disable(&timer);
+	cycle_ref = k_cycle_get_32();
```
Perhaps it would be good to use a a critical section for the above, so that there is no interruption between these operation? Similarly, for `nrfx_timer_capture()` with `k_cycle_get_32()` in `cpu_load_get()`.

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,234 @@
+	 * is possible that sleep time is bigger than total measured time.
+	 *
+	 * Note that for simplicity reasons module does not handle TIMER
+	 * overflow.
```
Perhaps it could be mentioned in the API description that a 32-bit register is used for counting microseconds of sleep time and thus `cpu_load_reset()` should be called from time to time (at least once per ~4294 seconds of sleeping)?

nordic-krch replied:
added.

anangl commented on include/debug/cpu_load.rst:
```diff
@@ -0,0 +1,65 @@
+    In the periodic load measurement logging, the :cpp:func:`cpu_load_get` function is called alternately with the :cpp:func:`cpu_load_reset`.
+
+Resetting the measurement
+    You can reset the the TIMER peripheral and the system clock by using :cpp:func:`cpu_load_reset`.
```
Resetting of the system clock sounds a bit scary. Maybe this sentence could be rephrased?

nordic-krch replied:
changed to `system clock readout`

nordic-krch commented on CODEOWNERS:
```diff
@@ -107,6 +107,8 @@ Kconfig*                                  @SebastianBoe
 /subsys/partition_manager/                @hakonfam
 /subsys/profiler/                         @pdunaj
+/subsys/debug/ppi_trace/                  @nordic-krch
+/subsys/debug/cpu_load/                   @nordic-krch
```
done

anangl replied:
I meant both those lines. First of them is still added by this PR.

nordic-krch commented on subsys/debug/cpu_load/Kconfig:
```diff
@@ -0,0 +1,84 @@
+	depends on !SOC_SERIES_NRF51X #Lack of required HW events
+	help
+	  Enable the CPU load measurement instrumentation. This tool is using
+	  one TIMER peripheral and PPI to perfrom accurate CPU load measurement.
```
seems like checkpatch is not run on nrf PRs. Is that right? fixed.

nordic-krch commented:
@greg-fer changes added here https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1272/commits/7634e0761b5182f48e9aaac149f96cc6e2c35bc8 can you review?

@anangl, thanks for review. please take another look. I've resolved trivial comments.

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,297 @@
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load / 1000;
+	u32_t fraction = 1000 % percent;
```
This is wrong. Should be `load % 1000`.
The same could be used in `cmd_cpu_load_get` as well.

nordic-krch replied:
my bad.

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,297 @@
+	int key = irq_lock();
+
+	nrfx_timer_enable(&timer);
+	nrfx_timer_disable(&timer);
```
I guess it should be the other way round, so that the timer stays enabled afterwards.

nordic-krch replied:
i missed your edit message but came to the same conclusion and replaced it with clearing. Also removed the lock because timer will not progress until CPU sleeps.

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,297 @@
+
+	key = irq_lock();
+	sleep_us = sleep_ticks_to_us(nrfx_timer_capture(&timer, 0));
+	total_cyc = k_cycle_get_32() - cycle_ref;
```
No need to perform the calculations with interrupts disabled, just covering readouts would be sufficient.

nordic-krch replied:
i actually came to the conclusion that locking is not needed. timer is measuring sleep time so it will not change even when context is interrupted.

greg-fer commented on include/debug/cpu_load.rst:
```diff
@@ -48,7 +48,7 @@ Getting the results
 
 Resetting the measurement
-    You can reset the the TIMER peripheral and the system clock by using :cpp:func:`cpu_load_reset`.
+    You can reset the TIMER peripheral and the system clock readout by using :cpp:func:`cpu_load_reset`.
```
```suggestion
    You can reset the TIMER peripheral and the system clock read-out by using :cpp:func:`cpu_load_reset`.
```
`read-out`, according to Merriam Webster dictionary.

nordic-krch commented:
@anangl please take a look at https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1272/commits/645f25427cd216ee98dfd7e6f44c1920663985cc which contains changes from latest review comments

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,289 @@
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
```
`ppi_cleanup()` missing here.

nordic-krch replied:
moved timer init before ppi alloc so that clean up is not needed.

anangl replied:
This way `CONFIG_CPU_LOAD_ALIGNED_CLOCKS` won't work.

nordic-krch replied:
done, moved init after handling `CONFIG_CPU_LOAD_ALIGNED_CLOCKS ` added there ppi cleanup

anangl replied:
Okay. If you want to keep it this way, you need to also call `nrfx_timer_uninit()` when the allocation of `ch_sleep` or `ch_wakeup` fails.

nordic-krch replied:
right, in that case it was better to extend ppi_cleanup to be able to cleanup wakeup channel, too.

nordic-krch commented:
@anangl please take a look at the latest commit with review comments

anangl commented on subsys/debug/cpu_load/cpu_load.c:
```diff
@@ -0,0 +1,292 @@
+	u32_t percent = load / 1000;
+	u32_t fraction = load % 1000;
+
+	if (!ready) {
```
Last one detail. It's a bit late for this check here. `cpu_load_get()` has been already called.

nordic-krch commented:
@rlubos @joerchan PR is green and has 3 approvals (including technical writer). Can you merge it? Applications team is waiting for it.

rlubos commented:
Changelog updated.
